File: core/agent-runtime/src/heartbeat.ts
/**
 * Heartbeat — sends periodic heartbeats to Core to prove the container is alive.
 */

import axios, { AxiosError } from 'axios';
import { log } from './logger.js';

const DEFAULT_INTERVAL_MS = 10_000; // 10 seconds

/**
 * Start sending heartbeats to Core.
 * Returns a cleanup function to stop the heartbeat interval.
 */
export function startHeartbeat(
  coreUrl: string,
  agentId: string,
  identityToken: string,
  intervalMs: number = DEFAULT_INTERVAL_MS,
): () => void {
  let running = true;

  const sendBeat = async () => {
    if (!running) return;

    try {
      await axios.post(
        `${coreUrl}/api/agents/${agentId}/heartbeat`,
        {},
        {
          headers: {
            'Authorization': `Bearer ${identityToken}`,
            'Content-Type': 'application/json',
          },
          timeout: 5000,
        },
      );
      log('debug', 'Heartbeat sent');
    } catch (err) {
      const msg = err instanceof AxiosError ? err.message : String(err);
      log('warn', `Heartbeat failed: ${msg}`);
      // Don't crash — heartbeat failures are non-fatal
    }
  };

  // Send first heartbeat immediately
  sendBeat();

  const timer = setInterval(sendBeat, intervalMs);

  return () => {
    running = false;
    clearInterval(timer);
    log('info', 'Heartbeat stopped');
  };
}

File: rust/crates/sera-runtime/src/heartbeat.rs
//! Heartbeat — periodic pings to sera-core to announce agent liveness.

use reqwest::Client;
use tracing::{debug, warn};

pub async fn run(
    core_url: String,
    instance_id: String,
    identity_token: String,
    interval_ms: u64,
) {
    let client = Client::new();
    let mut interval = tokio::time::interval(std::time::Duration::from_millis(interval_ms));

    loop {
        interval.tick().await;

        let url = format!("{}/api/agents/{}/heartbeat", core_url, instance_id);

        match client
            .post(&url)
            .header("Authorization", format!("Bearer {}", identity_token))
            .send()
            .await
        {
            Ok(resp) if resp.status().is_success() => {
                debug!("Heartbeat sent successfully");
            }
            Ok(resp) => {
                warn!("Heartbeat failed with status: {}", resp.status());
            }
            Err(e) => {
                warn!("Heartbeat error: {}", e);
            }
        }
    }
}
